File: core/src/main/java/art/arcane/iris/core/safeguard/task/Task.java
package art.arcane.iris.core.safeguard.task;

import art.arcane.iris.core.safeguard.Mode;

import java.util.function.Supplier;

public abstract class Task {
    private final String id;
    private final Mode failureMode;
    private final String failureLockReason;

    protected Task(String id, Mode failureMode, String failureLockReason) {
        this.id = id;
        this.failureMode = failureMode;
        this.failureLockReason = failureLockReason;
    }

    public String getId() {
        return id;
    }

    public Mode failureMode() {
        return failureMode;
    }

    public String failureLockReason() {
        return failureLockReason;
    }

    public abstract CheckResult run();

    public static Task advisory(String id, Supplier<CheckResult> action) {
        return new Task(id, Mode.WARNING, null) {
            @Override
            public CheckResult run() {
                return action.get();
            }
        };
    }

    public static Task critical(String id, String throwLockReason, Supplier<CheckResult> action) {
        return new Task(id, Mode.UNSTABLE, throwLockReason) {
            @Override
            public CheckResult run() {
                return action.get();
            }
        };
    }
}

File: core/src/main/kotlin/com/volmit/iris/core/safeguard/task/Task.kt
package com.volmit.iris.core.safeguard.task

import com.volmit.iris.core.safeguard.Mode
import com.volmit.iris.util.format.Form
import kotlin.properties.PropertyDelegateProvider
import kotlin.properties.ReadOnlyProperty

abstract class Task(
    val id: String,
    val name: String = Form.capitalizeWords(id.replace(" ", "_").lowercase()),
) {

    abstract fun run(): ValueWithDiagnostics<Mode>

    companion object {
        fun of(id: String, name: String = id, action: () -> ValueWithDiagnostics<Mode>) = object : Task(id, name) {
            override fun run() = action()
        }

        fun of(id: String, action: () -> ValueWithDiagnostics<Mode>) = object : Task(id) {
            override fun run() = action()
        }

        fun task(action: () -> ValueWithDiagnostics<Mode>) = PropertyDelegateProvider<Any?, ReadOnlyProperty<Any?, Task>> { _, _ ->
            ReadOnlyProperty { _, property -> of(property.name, action) }
        }
    }
}